flush/sink.go

```diff
@@ -0,0 +1,177 @@
+		ams.msgBufferSize = size
+	}
+}
+
+// WithAckBufferSize sets the ack channel buffer size.
+func WithAckBufferSize(size int) AsyncMessageSinkOption {
+	return func(ams *AsyncMessageSink) {
+		ams.ackBufferSize = size
+	}
+}
+
+// AckFunc is a callback function executed by `Run` for each message succesfully produced. The ack
+// counter is only incremented once this function returns. If no AckFunc is provided the ack counter
+// is always incremented.
+type AckFunc func(msg substrate.Message) error
+
+// AsyncMessageSink wraps substrate.AsyncMessageSink and provides an interface for interaction with
+// the underlying sink, as well as the capability to flush the message buffer.
+type AsyncMessageSink struct {
+	ctx           context.Context
+	cancel        context.CancelFunc
+	wait          func() error
+	sink          substrate.AsyncMessageSink
+	msgs          uint64
+	msgBufferSize int
+	msgCh         chan substrate.Message
+	acks          uint64
+	ackBufferSize int
+	ackCh         chan substrate.Message
+	ackFn         AckFunc
+}
+
+// NewAsyncMessageSink returns a pointer a new AsyncMessageSink.
+// See examples/flush/sink.go for example usage.
+func NewAsyncMessageSink(ctx context.Context, sink substrate.AsyncMessageSink, opts ...AsyncMessageSinkOption) *AsyncMessageSink {
+	ctx, cancel := context.WithCancel(ctx)
+
+	ams := &AsyncMessageSink{
+		ctx:           ctx,
+		cancel:        cancel,
+		sink:          sink,
+		msgBufferSize: defaultMsgBufferSize,
+		ackBufferSize: defaultAckBufferSize,
+	}
+
+	for _, opt := range opts {
+		opt(ams)
+	}
+
+	ams.msgCh = make(chan substrate.Message, ams.msgBufferSize)
+	ams.ackCh = make(chan substrate.Message, ams.ackBufferSize)
+
+	return ams
+}
+
+// Run initialises message publishing using the underlying sink and blocks until either an error
+// occurs or the constructor context is done. If an AckFunc is configured, Run will execute it for
+// each ack recieved. If an error is returned, the user should cancel the constructor context to
+// prevent `Flush` from blocking.
+func (ams *AsyncMessageSink) Run() error {
+	group, groupctx := rungroup.New(ams.ctx)
+	ams.wait = group.Wait
+
+	group.Go(func() error {
+		return ams.sink.PublishMessages(groupctx, ams.ackCh, ams.msgCh)
+	})
+
+	group.Go(func() error {
+		for {
+			select {
+			case <-groupctx.Done():
+				return nil
+			case msg, ok := <-ams.ackCh:
+				if !ok {
+					return nil
+				}
+
+				if ams.ackFn != nil {
+					err := ams.ackFn(msg)
+					if err != nil {
+						return err
+					}
+				}
+
+				atomic.AddUint64(&ams.acks, 1)
+			}
+		}
+	})
+
+	return ams.wait()
+}
+
+// PublishMessage publishes a message to the underlying sink. PublishMessage is desgined to be
+// called concurrently by the user. The ctx passed to PublishMessage controls only the publishing
+// of the message and is a seperate concern to the constructor context.
+func (ams *AsyncMessageSink) PublishMessage(ctx context.Context, msg []byte) error {
+	select {
```

Comment:
This should also wait on to stop trying to write a message when the underlying sink or ack function errors 
```suggestion
	select {
	case <-ams.ctx.Done():
 		return ctx.Err()
```

Reply:
We have a test case for this scenario in which if you call `PublishMessage` after calling `Close` you will always get a panic. Because select evaluates cases concurrently with this change there's a chance an error could be returned or the code could panic. IMO I think more predicable code is better.

Reply:
this is the test case https://github.com/uw-labs/substrate-tools/pull/4/files#diff-3b4aba438d02a5aa8b1b0e87c78b181eR202-R229

Reply:
Ah missed that you are closing the messages channel, what is the reason for doing that? I don't think current substrate backends are expecting that atm, so it might lead to unexpected behaviour.  IMO it would be better to just return an error instead of panicking. That way you would also get a consistent behaviour...

Reply:
you can't get consistent behaviour if you leave the channel open:

i.e. select evaluates concurrently, if channel closed and msgCh is still open this code can endup either error or succeed with nil error

```
select {
case <-ams.ctx.Done():
 return ctx.Err()
case ams.msgCh <- message.NewMessage(msg):

}
```

as select does this concurrently, it's not a switch statement

Reply:
IMO this behaviour:
sink.Close()

sink.PublishMEssage() - should always panic

Reply:
If nothing is reading from the channel and the channel is not buffered, then the code will always end up in the context cancellation branch as the message send is blocked. 

If you want to support buffered channels and make sure you get consistent behaviour after calling close on the sink, you can first check if the context is done and only then do the select. That is do something like this
```
if ams.ctx.Err() != nil {
    return ams.ctx.Err()
}
select {
case <-ams.ctx.Done():
 	return ctx.Err()
case <-ctx.Done():
 	return ctx.Err()
case ams.msgCh <- message.NewMessage(msg):
 	atomic.AddUint64(&ams.msgs, 1)
 }
return nil
```

Reply:
> IMO this behaviour:
> 
> ```
> sink.Close()
> sink.PublishMessage(msg)
> ```
> 
> should always panic on PublishMessage, which it does right now as our test shows

I don't think panic is necessary in this case, especially as the function returns an error. The synchronous sink adapter just returns an error in this case https://github.com/uw-labs/substrate/blob/master/sync_adapter_sink.go#L151-L152. The content of panic atm is also not super helpful as it's about send on a closed channel and not sending a message after closing a substrate backend.

Reply:
Agree sending err >  panic. I actually prefer your code :+1:  thanks :bow:

Reply:
You're right. Returning is preferable to panicing. `PublishMessage` now returns `substrate.ErrSinkAlreadyClosed` when either context is complete